File: Server/src/GameMode.ts
import { GameMode } from "./actions.js"

type GameModeData = {
  startingLives: number

  getBlindFromAnte: (ante: number, options: any) => { 
    small?: string
    big?: string
    boss?: string 
  }

  // TODO: Validate lobby options when they differ per gamemode
}

const GameModes: {
  [key in GameMode]: GameModeData
} = {
  'attrition': {
    startingLives: 4,
    getBlindFromAnte: (ante, options) => {
      return { boss: 'bl_pvp' }
    }
  },
  'draft': {
    startingLives: 2,
    getBlindFromAnte: (ante, options) => {
      const starting_antes = options?.draft_starting_antes ? parseInt(options.draft_starting_antes) : 3
      if (ante <= starting_antes) return { }
      else return { small: 'bl_pvp', big: 'bl_pvp', boss: 'bl_pvp' }
    }
  }
}

export default GameModes

File: RustServer/src/game_mode.rs
use crate::lobby::GameMode;
use dashmap::DashMap;
use serde::{Deserialize, Serialize};
use tokio::sync::OnceCell;

#[derive(Serialize, Deserialize, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct GameInfo {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub small: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub big: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub boss: Option<String>,
}

#[derive(Debug, PartialEq)]
pub struct GameModeData {
    pub starting_lives: i32,
    // Takes ante and options, returns GameInfo
    pub get_blind_from_ante: fn(i32, DashMap<String, String>) -> GameInfo,
}

pub static GAME_MODES: OnceCell<DashMap<GameMode, GameModeData>> = OnceCell::const_new();

pub async fn initialize_gamemodes() -> DashMap<GameMode, GameModeData> {
    let gamemodes = DashMap::new();
    gamemodes.insert(
        GameMode::Attrition,
        GameModeData {
            starting_lives: 4,
            get_blind_from_ante: |_ante, _options| GameInfo {
                small: None,
                big: None,
                boss: Some("bl_pvp".to_string()),
            },
        },
    );
    gamemodes.insert(
        GameMode::Draft,
        GameModeData {
            starting_lives: 2,
            get_blind_from_ante: |_ante, _options| GameInfo {
                small: Some("bl_pvp".to_string()),
                big: Some("bl_pvp".to_string()),
                boss: Some("bl_pvp".to_string()),
            },
        },
    );

    gamemodes
}
